llama-index-core/llama_index/core/workflow/checkpointer.py

```diff
@@ -0,0 +1,180 @@
+import asyncio
+import uuid
+import functools
+from llama_index.core.bridge.pydantic import (
+    BaseModel,
+    Field,
+    ConfigDict,
+)
+from typing import (
+    Optional,
+    Dict,
+    Any,
+    List,
+    Protocol,
+    TYPE_CHECKING,
+    Type,
+    Awaitable,
+)
+from llama_index.core.workflow.context import Context
+from llama_index.core.workflow.context_serializers import BaseSerializer, JsonSerializer
+from llama_index.core.workflow.handler import WorkflowHandler
+from .events import Event
+from .errors import *
+
+if TYPE_CHECKING:  # pragma: no cover
+    from .workflow import Workflow
+
+
+class CheckpointCallback(Protocol):
+    def __call__(
+        self,
+        last_completed_step: Optional[str],
+        input_ev: Optional[Event],
+        output_ev: Event,
+        ctx: Context,
+    ) -> Awaitable[None]:
+        ...
+
+
+class Checkpoint(BaseModel):
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+    id_: str = Field(default_factory=lambda: str(uuid.uuid4()))
+    last_completed_step: Optional[str]
+    input_event: Optional[Event]
+    output_event: Event
+    ctx_state: Dict[str, Any]
+
+
+class WorkflowCheckpointer:
+    def __init__(
+        self,
+        workflow: "Workflow",
+        checkpoint_serializer: Optional[BaseSerializer] = None,
+        disabled_steps: List[str] = [],
+    ):
+        self._checkpoints: Dict[str, List[Checkpoint]] = {}
+        self._checkpoint_serializer = checkpoint_serializer or JsonSerializer()
+        self._lock: asyncio.Lock = asyncio.Lock()
+
+        self.workflow = workflow
+        self.checkpoints_config: Dict[str, bool] = {
```

Comment:
nit: This could use a better name. `is_step_checkpointed` ? `enabled_checkpoints` ?

Reply:
Should this just be a set instead of a dict?

Reply:
yeah i think a set is better, if we just go with showing what's enabled -- i'll make that change